ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Handlers.hs

```diff
@@ -144,3 +146,49 @@ handlerRequestNext currentIntersection blockTree points =
     TipPoint tip' = tip points
 
     trace = tell . pure
+
+handlerBlockFetch ::
+  forall m .
+  IOLike m =>
+  BlockTree TestBlock ->
+  ChainRange (Point TestBlock) ->
+  AdvertisedPoints ->
+  STM m (Maybe BlockFetch, [String])
+handlerBlockFetch blockTree (ChainRange from to) points =
+  runWriterT $ case points of
+    AdvertisedPoints {header = HeaderPoint (At h), block = BlockPoint (At b)} -> do
+      case BT.findFragment (headerPoint h) blockTree of
+        Just f  -> withHpFragment (f, b)
+        Nothing -> error "header point is not in the block tree"
+    AdvertisedPoints {} -> pure Nothing
+  where
+
+    withHpFragment (hpChain, bp) =
+      -- First, check whether the block point is on the same chain, and before or equal to, the header point.
+      case AF.rollback (blockPoint bp) hpChain of
+        Nothing ->
+          -- REVIEW: Should this be fatal?
+          error "Block point isn't on the same chain as header point"
+          -- pure (Just NoBlocks)
+        Just bpChain ->
+          -- Next, check whether the requested range is contained in the fragment before the block point.
+          -- REVIEW: this is Nothing if only _part_ of the range is on the current chain.
+          -- Is it correct that we don't want to send anything in this case?
```

Comment:
So right now we either send all the required blocks or nothing? I suppose this makes sense between honest nodes, though, because an honest client will only ask for blocks that you do indeed have and you will always serve all of them. Calling @amesgen or @nfrisby though; they should know better.

Reply:
I think this was clarified on [Slack](https://moduscreate.slack.com/archives/C05CEUFFH7Z/p1701272848904619). Answer: yes, it is correct to send `MsgNoBlocks` unless all blocks are available.

Reply:
clarified on slack, updated the comment